File: data_structure/binaryTree/preorder/preoder_traversal_recursion.go
package main

import (
	"fmt"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func main() {

	root := &TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 2}}
	root.Left.Left = &TreeNode{Val: 3}
	root.Left.Right = &TreeNode{Val: 4}
	root.Right.Left = &TreeNode{Val: 4}
	root.Right.Right = &TreeNode{Val: 3}
	result := preorderTraversal(root)
	fmt.Println(result)

}

// 递归 前序遍历二叉树
func preorderTraversal(root *TreeNode) []int {
	result := []int{}
	if root == nil {
		return result
	}
	result = append(result, root.Val)

	left := preorderTraversal(root.Left)
	result = append(result, left...)

	right := preorderTraversal(root.Right)
	result = append(result, right...)

	return result
}

File: data_structure/binaryTree/preorder/preoder_traversal_recursion.py

# 二叉树的前序遍历

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# 递归
def preorderTraversal(root: TreeNode) ->[int]:
    result = []
    if not root:
        return result
    result.append(root.val)
    left = preorderTraversal(root.left)
    result.extend(left)
    right = preorderTraversal(root.right)
    result.extend(right)
    return result



if __name__ == "__main__":
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(2)

    root.left.left = TreeNode(3)
    root.left.right = TreeNode(4)

    root.right.left = TreeNode(4)
    root.right.right = TreeNode(3)
    result = preorderTraversal(root)
    print(result)